lib/api.js

```diff
@@ -48,6 +49,40 @@ function removeCoupling(heroku, app) {
            .then(coupling => deleteCoupling(heroku, coupling.id));
 }
 
+function listCouplings(heroku, pipelineId) {
+  return heroku.request({
+    method: 'GET',
+    path: `/pipelines/${pipelineId}/pipeline-couplings`,
+    headers: { 'Accept': V3_HEADER }
+  });
+}
+
+function getAppFilter(heroku, appIds) {
+  return heroku.request({
+    method: 'POST',
+    path: `/filters/apps`,
+    headers: { 'Accept': FILTERS_HEADER },
+    body: { in: { id: appIds } }
+  });
+}
+
+function listPipelineApps(heroku, pipelineId) {
+  return listCouplings(heroku, pipelineId).then((couplings) => {
+    const appIds = couplings.map((coupling) => coupling.app.id);
+
+    return getAppFilter(heroku, appIds).then((apps) => {
+      const couplingsByAppId = couplings.reduce((memo, coupling) => {
```

Comment:
Can you extract a `groupBy` helper here? Reductions can be hard to parse and it'll simplify the code.